### PR TITLE
LPS-166112 Add support to retrieve nestedFields with the relationship name

### PR DESCRIPTION
**Pull request by LuismiBarcos**
Related tickets 
* Sub-task: [LPS-166112](https://issues.liferay.com/browse/LPS-166112)
* Story: [LPS-161364](https://issues.liferay.com/browse/LPS-161364)
____
## Purpose of this PR
Add support to retrieve the related object that comes through nested fields using the relationship name as a query parameter in nested fields. This is behavior supported by the many-to-many relationships but not supported by the one-to-many.

### Before the PR
If a user makes the following request
`curl "http://localhost:8080/o/c/students/41669"`
The client received:
```
{
... 
  "studentName": "Student 1",
  "r_universityStudents_c_universityId": 41665
}
```
And, if the client used `r_universityStudents_c_university`
`curl "http://localhost:8080/o/c/students/41669?nestedFields=r_universityStudents_c_universityId"`
will receive
```
...
 "r_universityStudents_c_university": {
    "creator": {
      "additionalName": "",
      "contentType": "UserAccount",
      "familyName": "Test",
      "givenName": "Test",
      "id": 20125,
      "name": "Test Test"
    },
    "dateCreated": "2022-10-13T17:06:36Z",
    "dateModified": "2022-10-13T17:06:36Z",
    "externalReferenceCode": "41665",
    "id": 41665,
    "status": {
      "code": 0,
      "label": "approved",
      "label_i18n": "Approved"
    },
    "universityName": "Oxford"
  },
  "studentName": "Student 1",
  "r_universityStudents_c_universityId": 41665
}
```
Note that `r_universityStudents_c_university` is the related object (a university in this example)

### After the PR
For this request
`curl "http://localhost:8080/o/c/students/41669"`

will receive
```
{
...
  "universityStudentsId": 41665,
  "studentName": "Student 1",
  "r_universityStudents_c_universityId": 41665
}
```
A new field has come: `universityStudentsId` which is exactly the same as `r_universityStudents_c_universityId` but now is the relationship name, making it clearer.

Of course, a client could do the following request
`curl "http://localhost:8080/o/c/students/41669?nestedFields=universityStudents" `
and will receive
```
...
  "universityStudents": {
    "creator": {
      "additionalName": "",
      "contentType": "UserAccount",
      "familyName": "Test",
      "givenName": "Test",
      "id": 20125,
      "name": "Test Test"
    },
    "dateCreated": "2022-10-13T17:06:36Z",
    "dateModified": "2022-10-13T17:06:36Z",
    "externalReferenceCode": "41665",
    "id": 41665,
    "status": {
      "code": 0,
      "label": "approved",
      "label_i18n": "Approved"
    },
    "universityName": "Oxford"
  },
  "studentName": "Student 1",
  "r_universityStudents_c_universityId": 41665
}
```
Exactly the same related object but with the relationship name.

Of course, the approach followed before the PR is supported and can be used. This is to avoid breaking changes.

## How to test it
* Deploy `object-rest-impl`
* Enable the following feature flag: `feature.flag.LPS-161364`

### CODE REVIEW

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by LuismiBarcos** [comment]
ci:reevaluate:1297401_5044

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java [hunk 1]
```diff
@@ -121,6 +121,29 @@ public ObjectEntry toDTO(
 			objectEntry);
 	}
 
+	private void _addNestedFields(
+		Map<String, Object> map, String objectFieldName,
+		ObjectRelationship objectRelationship, String nestedFields,
```
Comment (4lejandrito) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java:127]: Sort the params alphabetically.

**Review by 4lejandrito — CHANGES_REQUESTED** [review]

**Comment by LuismiBarcos** [comment]
CI, I hate you. Last try before contact with QA to see if it is safe to send it manually.

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by magjed4289** [comment]
Just started reviewing :)

**Comment by magjed4289** [comment]
Failures that seem to be related:
[LocalFile.SystemObjectAccount#CanAssociateAnyTypeOfAccount](https://testray.liferay.com/home/-/testray/case_results/2008267075?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2008266101%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)

[com.liferay.object.rest.internal.manager.v1_0.test.DefaultObjectEntryManagerImplTest
](https://testray.liferay.com/home/-/testray/case_results/2008269148/history)

Test fix will be provided for:
LocalFile.AdvancedObjectAPI#CanReturnNestedFieldsDetailsInOneToManyUpdatedRelationship
LocalFile.AdvancedObjectAPI#CanReturnNestedFieldsDetailsInOneToManyUpdatedRelationship

**Comment by magjed4289** [comment]
ci:test:headless-functional

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by LuismiBarcos** [comment]
After some issues that appears developing this, here is the final PR (I hope). See this [slack](https://liferay.slack.com/archives/C03U66A6NNN/p1666188758928739) conversation to have more context about the first commit.

I also add some tests and do some refactor in the tests code, that's why I added @ccorreagg as a reviewer.

**Comment by ccorreagg** [comment]
I think the description contains some errors:

Not sure if the nestedField must be r_universityStudents_c_university or r_universityStudents_c_universityId
```
And, if the client used r_universityStudents_c_university
curl "http://localhost:8080/o/c/students/41669?nestedFields=r_universityStudents_c_universityId"
will receive
```

```
{
...
  "universityStudents": 41665,
  "studentName": "Student 1",
  "r_universityStudents_c_universityId": 41665
}
A new field has come: universityStudentsId which is exactly the same as r_universityStudents_c_universityId but now is the relationship name, making it clearer.

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java [hunk 2]
```diff
@@ -121,6 +121,29 @@ public ObjectEntry toDTO(
...
+
+		String[] nestedFieldsList = nestedFields.split(",");
+
+		String objectFieldNameNestedField = StringUtil.replaceLast(
+			objectFieldName.substring(
+				objectFieldName.lastIndexOf(StringPool.UNDERLINE) + 1),
+			"Id", "");
+
+		for (String nestedField : nestedFieldsList) {
+			if (nestedField.contains(objectFieldNameNestedField)) {
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java:137]: Are we keeping the old "magic" functionality then?

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+
+		Assert.assertEquals(1, itemsJSONArray.length());
+
+		JSONObject itemJSONObject = itemsJSONArray.getJSONObject(0);
+
+		Assert.assertEquals(
+			relatedObjectFieldValue,
+			itemJSONObject.getString(relatedObjectFieldName));
+
+		_invoke(
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java]: I don't know if this invocation makes sense as you are not getting the result of the GET request

**Review by ccorreagg — COMMENTED** [review]
A few comments. Anyway, I would wait until PM decission as I think it is a bit tricky for the future OpenAPI generation

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+			objectDefinition.getObjectDefinitionId());
+	}
+
+	private static final String _OBJECT_FIELD_NAME =
+		"x" + RandomTestUtil.randomString();
+
+	private static final String _OBJECT_FIELD_VALUE =
+		RandomTestUtil.randomString();
+
+	@DeleteAfterTestRun
```
Comment (ccorreagg) [inline modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:169]: I suspect if we don't delete the object definition manually, the created tables in execution time will be alive forever as I think the Rule deletes only the row in the table. Could you test it?

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java [hunk 5]
```diff
@@ -403,57 +431,43 @@ else if ((nestedFieldsDepth > 0) &&
...
 									nestedFieldsDepth - 1,
 									_objectEntryLocalService.getObjectEntry(
-										objectEntryId)));
+										objectEntryId)),
+								objectFieldName, objectRelationship);
 						}
 					}
 				}
 
+				map.putIfAbsent(objectRelationship.getName(), objectEntryId);
```
[the remark this answers is not in the document]
Comment (LuismiBarcos) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java]: I am changing it as our PM decide to change the name of the field

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java (hunk 2 above)
Comment (LuismiBarcos) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java:137]: Well, as you know, we decide to keep the old "magical" functionality

File: modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java (hunk 4 above)
Comment (LuismiBarcos) [reply modules/apps/object/object-rest-test/src/testIntegration/java/com/liferay/object/rest/internal/resource/v1_0/test/ObjectEntryResourceTest.java:169]: I tested it and it is ok. The annotation `@DeleteAfterTestRun` calls the `deleteObjectDefinition` method that deletes also the table

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by LuismiBarcos** [comment]
ci:test:relevant

File: modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java (hunk 2 above)
Comment (ccorreagg) [reply modules/apps/object/object-rest-impl/src/main/java/com/liferay/object/rest/internal/dto/v1_0/converter/ObjectEntryDTOConverter.java:137]: :face_in_clouds:

**Review by ccorreagg — APPROVED** [review]
LGTM!

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by ccorreagg** [comment]
ci:test:relevant

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by LuismiBarcos** [comment]
ci:forward

**Comment by Hijky** [comment]
Hi @LuismiBarcos  I think the test failures are unrelated to your pr. Here is the [compare run](https://testray.liferay.com/home/-/testray/runs/compare?compare=true&view=details&statusA=3&testrayRunIdB=2021035134&testrayRunIdA=2020138400&statusB=2) results:

integration: com.liferay.object.service.test.ObjectActionLocalServiceTest -> also happen on [other Acceptance](https://testray.liferay.com/home/-/testray/case_results/2020427487) test
com.liferay.sharing.search.internal.test.SharingEntrySearchPermissionCheckerTest ->It seems failed before run test
And the other are not related with PR.

**Comment by LuismiBarcos** [comment]
Sent here https://github.com/brianchandotcom/liferay-portal/pull/125440

**Pull request closed without merging**